File: src/io/ivy/fawkes/Events.java
// -*- Mode: jde; eval: (hs-hide-level 2) -*-

package io.ivy.fawkes;

import java.util.List;

import org.bukkit.*;
import org.bukkit.block.*;
import org.bukkit.metadata.*;
import org.bukkit.entity.*;
import org.bukkit.event.player.*;
import org.bukkit.event.block.*;
import org.bukkit.event.entity.*;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;

import org.bukkit.inventory.ItemStack;

public class Events implements Listener {

  private final Fawkes fawkes;

  public Events(Fawkes instance) {
    fawkes = instance;
  }

  @EventHandler
  public void onPlayerInteract(PlayerInteractEvent event) {
    if (event.getPlayer().hasPermission("fawkes.fks")) {
      if (event.getAction().equals(Action.RIGHT_CLICK_BLOCK)) {
        if (event.getClickedBlock().getType().equals(Material.CHEST)) {
          if (event.getPlayer().getItemInHand() != null) {
            
            ItemStack held = event.getPlayer().getItemInHand();
            Chest chest = (Chest) event.getClickedBlock().getState();
            
            if (held.getType().equals(Material.STICK)) {

              String loot_type = held.getItemMeta().getLore().get(0);

              if (loot_type.equals("regular")) {
                mark_chest_regular(chest);
              }
              
              if (loot_type.equals("large")) {
                mark_chest_large(chest);
              }

              if (loot_type.equals("murca")) {
                mark_chest_murca(chest);
              }
              
              if (loot_type.equals("view")) {
            	  fawkes.log("Loot Tag: " + chest.getMetadata("ivy.loot"));
              }
            }
          }
        }
      }
    }
  }

  @EventHandler
  public void onEntityDamageByEntity(EntityDamageByEntityEvent event) {
    
    Entity damager = event.getDamager();
    Entity target = event.getEntity();
    
    Double pre_damage = Double.valueOf(event.getDamage());
    Double end_damage = Double.valueOf(event.getFinalDamage());
    
    String damager_type = damager.getType().toString();
    String target_type = target.getType().toString();
    
    int damager_level = find_mob_level(damager);
    int target_level = find_mob_level(target);
    
    double level_difference = damager_level - target_level;
    
    double new_damage_level = end_damage + (level_difference * .5);
    
    if (new_damage_level < 0) {
      new_damage_level = 0;
    }
    
    fawkes.log(damager + "(" + find_mob_level(damager) + ") HIT " +
               target + "(" + find_mob_level(target) + ") for " +
               pre_damage + "/" + end_damage + " -> " + new_damage_level);
    event.setDamage(new_damage_level);
  }
  
  @EventHandler
  public void onEntityDeath(EntityDeathEvent event) {
    
    if (event.getEntity().getKiller() != null) {
      
      Entity entity = event.getEntity();
      Player player = event.getEntity().getKiller();
      
      int player_level = player.getLevel();
      int mob_level = find_mob_level(entity);

      int xp_to_add = 0;
      
      if (player_level > mob_level) {
        if ((player_level - mob_level) > 4) {
          return;
        }
      }

      if (mob_level <= 5) {
        xp_to_add = 1;
      }
      if (mob_level > 5 && mob_level <= 9) {
        xp_to_add = 2;
      }

      if (mob_level > 9 && mob_level <= 17) {
        xp_to_add = 3;
      }

      if (mob_level > 17 && mob_level <= 28) {
        xp_to_add = 5;
      }
      
      player.giveExp(xp_to_add);
      player.sendMessage("You recieve " + xp_to_add + " experience!");

    }
  }
  
  @EventHandler
  public void onPlayerDeath(PlayerDeathEvent event) {
  }
  
  @EventHandler
  public void onFoodLevelChange(FoodLevelChangeEvent event) {
    
    if (event.getEntityType().equals(EntityType.PLAYER)) {
      if (event.getFoodLevel() < 10) {
        Player player = (Player) event.getEntity();
        int food_level = event.getFoodLevel();
        fawkes.log("Food level: " + String.valueOf(food_level));
      }
    }
  }

  

  private int find_mob_level(Entity entity) {
    
    if (entity.hasMetadata("NPC")) {
      return 5;
    }
    
    if (entity.getType().equals(EntityType.LIGHTNING)) {
      return 5;
    }
        
    if (entity.getType().equals(EntityType.PLAYER)) {
      return ((Player)entity).getLevel();
    }
    
    if (entity instanceof Projectile) {
      Projectile p = (Projectile) entity;
      if (p.getShooter() instanceof Player) {
        return find_mob_level((Player)p.getShooter());
      }
    }
    
    if (entity instanceof LightningStrike) {
      fawkes.log("BZZZZZT");
    }
    
    List<MetadataValue> level_values = entity.getMetadata("ivy.level");
    
    if (level_values.isEmpty()) {
      return 1;
    } else {
      return level_values.get(0).asInt();
    }
  }

  public int find_zero_difference(int cl) {
    // cl = character level.
    
    if (cl <= 7)
      return 5;
    if (cl >= 8 && cl < 9)
      return 6;
    if (cl >= 10 && cl <= 11)
      return 7;
    if (cl >= 12 && cl <= 15)
      return 8;
    
    if (cl >= 16 && cl <= 19)
      return 9;
    if (cl >= 20 && cl <= 29)
      return 11;
    if (cl >= 30 && cl <= 39)
      return 12;
    if (cl >= 40 && cl <= 44)
      return 13;
    if (cl >= 45 && cl <= 49)
      return 14;
    if (cl >= 50 && cl <= 54)
      return 15;
    if (cl >= 55 && cl <= 59)
      return 16;
    if (cl >= 60 && cl <= 79)
      return 17;
    return 18;
  }

  private void mark_chest_regular(Chest chest) {
	  chest.setMetadata("ivy.loot", new FixedMetadataValue(fawkes, "small"));
    fawkes.log("Little loots.");
  }

  private void mark_chest_large(Chest chest) {
	  chest.setMetadata("ivy.loot", new FixedMetadataValue(fawkes, "medium"));
    fawkes.log("Kinda big loots.");
  }

  private void mark_chest_murca(Chest chest) {
	  chest.setMetadata("ivy.loot", new FixedMetadataValue(fawkes, "murca"));
    fawkes.log("Fat ass fuckin' loots!");
  }

}

File: src/ivy/fawkes/events.clj
;; -*- Mode: Clojure; eval: (hs-hide-all) -*-

(ns ivy.fawkes.events
  
  (:require [ivy.fawkes.bukkit.event :as event]
            [ivy.fawkes.block :as block]
            [ivy.fawkes.util :as u]
            [monger.core :as mg]
            [monger.collection :as mc])
  
  (:import [org.bukkit.block Chest Biome]
           [org.bukkit.entity Entity EntityType Projectile Player Monster]
           [org.bukkit Material Bukkit ChatColor]
           [org.bukkit.event EventPriority]
           [org.bukkit.event.player PlayerInteractEvent PlayerRespawnEvent]
           [org.bukkit.event.block Action]
           [org.bukkit.event.world ChunkLoadEvent]
           [org.bukkit.event.entity EntityDamageByEntityEvent EntityDamageEvent CreatureSpawnEvent EntityDeathEvent]
           [org.bukkit.metadata FixedMetadataValue]
           [org.bukkit.inventory ItemStack]))

(defonce ^:dynamic fawkes (atom nil))
(defonce ^:dynamic mongo (atom nil))

(defn on-player-interact [^PlayerInteractEvent event]
  (let [player (.getPlayer event)
        block (.getClickedBlock event)
        in-hand (.getItem event)]
    (when (and in-hand block)
      (when (and (=
                  (.equals (.getType block) Material/CHEST)
                  (.equals (.getType in-hand) Material/STICK)))
        
        (when (.getLore (.getItemMeta in-hand))
          (let [lore (.get (.getLore (.getItemMeta in-hand)) 0)]
            
            (when (= lore "view")
              (let [metadata (.getMetadata (.getState block) "ivy.loot")]
                (if (> (.size metadata) 0)
                  (.sendMessage player (format "Metadata: %s" (.value (.get metadata 0))))
                  (.sendMessage player "No metadata!"))))
            
            (when (= lore "regular")
              (block/save-block block "ivy.loot" "regular")
              (.setMetadata block "ivy.loot" (new FixedMetadataValue @fawkes "regular"))
              (.sendMessage player "Loot type set."))
            
            (when (= lore "large")
              (block/save-block block "ivy.loot" "large")
              (.setMetadata block "ivy.loot" (new FixedMetadataValue @fawkes "large"))
              (.sendMessage player "Loot type set."))
            
            (when (= lore "murca")
              (block/save-block block "ivy.loot" "murca")
              (.setMetadata block "ivy.loot" (new FixedMetadataValue @fawkes "murca"))
              (.sendMessage player "Loot type set."))))))))

(defn on-chunk-load [event]
  (let [chunk (.getChunk event)]
    (doseq [tentity (.getTileEntities chunk)]
      (.info (.getLogger @fawkes) "Tile Entity...")
      (when (.equals (.getType tentity) Material/DIAMOND_ORE)
        (.info (.getLogger @fawkes) "Diamond ore."))
      (when (.equals (.getType tentity) Material/REDSTONE_ORE)
        (.info (.getLogger @fawkes) "Redstone ore."))
      (when (.equals (.getType tentity) Material/LAPIS_ORE)
        (.info (.getLogger @fawkes) "Lapis ore located.")
        true))))

;; (defn on-player-respawn [event]
;;   (let [player (.getPlayer event)
;;         name (.getName player)
;;         dead-level (wcar* (car/get (format "fawkes.death.%s.level" name)))]
;;     (when dead-level
;;       (wcar* (car/del (format "fawkes.death.%s.level" name)))
;;       (.info (.getLogger @fawkes) (format "Restoring level to %s from death event." name))
;;       (.setLevel player (u/parse-int dead-level)))))

(defn find-mob-level [^Entity entity]
  (cond (.hasMetadata entity "NPC") 50
        (instance? Projectile entity) (when (instance? Player (.getShooter entity))
                                        (find-mob-level (.getShooter entity)))
        (instance? Monster entity) (let [level (.getMetadata entity "ivy.level")]
                                     (if (.isEmpty level)
                                       1
                                       (.. level (get 0) (asInt))))
        
        (instance? Player entity) (or (.getLevel entity) 1)
        :else 1))

(defn on-entity-death [^EntityDeathEvent event]
  (let [killed (.getEntity event)]

    
    ;; (when (instance? Player killed)
    ;;   (let [the-level (.getLevel killed)]
    ;;     (.info (.getLogger @fawkes) (format "Player killed - saving level."))
    ;;     (wcar* (car/set (format "fawkes.death.%s.level" (.getName killed)) (- the-level 1)))))

    (let [killed (.getEntity event)
          killer (.getKiller killed)]
      (when (and (instance? Monster killed)
                 (instance? Player killer))
        
        (let [killed-level (find-mob-level killed)
              killer-level (.getLevel killer)]
          
          (when (<= killed-level 5)
            (u/give-exp killer 1))
          (when (and (> killed-level 5) (<= killed-level 9))
            (u/give-exp killer 2))
          (when (and (> killed-level 9) (<= killed-level 17))
            (u/give-exp killer 3))
          (when (and (> killed-level 17) (<= killed-level 28))
            (u/give-exp killer 5)))))))

(defn on-entity-damage [^EntityDamageByEntityEvent event]
  (when (instance? EntityDamageByEntityEvent event)
    (let [attacker (.getDamager event)
          defender (.getEntity event)
          damage (.getDamage event)
          final-damage (.getFinalDamage event)
          attacker-level (find-mob-level attacker)
          defender-level (find-mob-level defender)
          level-difference (- attacker-level defender-level)
          t-new-damage (+ final-damage (* level-difference 0.5))
          new-damage (cond (<= t-new-damage 0) 1
                           :else t-new-damage)]
      (.info (.getLogger @fawkes) (format "%s (%s) HIT %s (%s) for %s/%s -> %s"
                                          attacker attacker-level
                                          defender defender-level
                                          damage final-damage
                                          new-damage))
      (.setDamage event new-damage))))

(defn spawn-creature [level entity biome]
  (.setCustomName entity (format "§e(%s) %s" level (.getType entity)))
  (.setCustomNameVisible entity true)
  (.setMetadata entity "ivy.level" (FixedMetadataValue. @fawkes level)))

(defn on-entity-spawn [event]
  (let [entity (.getEntity event)
        location (.getLocation entity)
        world (.getWorld location)
        loc-x (.getBlockX location)
        loc-y (.getBlockY location)
        biome (.getBiome world loc-x loc-y)
        high-y (.getHighestBlockYAt world location)]
    (do
      (when (= (.getType (.getEntity event)) EntityType/RABBIT)
        (spawn-creature 50 (.getEntity event) biome))
      (when (and (instance? CreatureSpawnEvent event)
                 (instance? Monster (.getEntity event)))
        (let [location (.getLocation (.getEntity event))
              world (.getWorld location)
              loc-x (.getBlockX location)
              loc-y (.getBlockY location)
              biome (.getBiome world loc-x loc-y)
              entity (.getEntity event)]
          (cond
            
            (= biome Biome/FOREST) (spawn-creature (u/rand-range 4 8) entity biome)
            (= biome Biome/JUNGLE) (spawn-creature (u/rand-range 3 5) entity biome)
            (= biome Biome/JUNGLE_HILLS) (spawn-creature (u/rand-range 2 4) entity biome)
            (= biome Biome/PLAINS) (spawn-creature (u/rand-range 8 15) entity biome)
            (= biome Biome/TAIGA) (spawn-creature (u/rand-range 8 15) entity biome)
            (= biome Biome/RIVER) (spawn-creature (u/rand-range 3 7) entity biome)
            
            :else (do
                    (u/info "Request for biome %s fell through." biome)
                    (spawn-creature 1 entity biome))))))))

(defn handle-event [f e]
  (if-let [response (f e)]
    (do
      (if (:msg response)
        (.sendMessage e response)))))

(defn start [plugin]
  (reset! fawkes plugin)
  (reset! mongo (mg/connect))
 
  (u/info "Loading events...")

  (event/register-event plugin "org.bukkit.event.player.PlayerInteractEvent" #'on-player-interact)
  ;(event/register-event plugin "org.bukkit.event.player.PlayerRespawnEvent" #'on-player-respawn)
  (event/register-event plugin "org.bukkit.event.entity.EntityDeathEvent" #'on-entity-death)
  (event/register-event plugin "org.bukkit.event.entity.EntityDamageByEntityEvent" #'on-entity-damage)
  (event/register-event plugin "org.bukkit.event.entity.CreatureSpawnEvent" #'on-entity-spawn))
